File: algorithm/19_2.cpp
#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x=0): val(x), next(nullptr) {}
};

class Solution
{
public:
    ListNode* removeNthFromEnd(ListNode* head, int n);
};

ListNode* Solution::removeNthFromEnd(ListNode* head, int n)
{
    ListNode* dummyhead = new ListNode(0);
    dummyhead->next = head;
    ListNode* prev = dummyhead;
    ListNode* fast = head;
    ListNode* slow = head;

    for (int i = 0; i < n; i++)
    {
        if (fast==nullptr)
        {
            return head;
        }
        fast = fast->next;
    }
    
    while (fast!=nullptr)
    {
        prev = slow;
        slow = slow->next;
        fast = fast->next;
    }
    prev->next = slow->next;
    head = dummyhead->next;
    delete dummyhead;
    return head;
}

int main()
{
    ListNode* n1 = new ListNode(1);
    ListNode* n2 = new ListNode(2);
    ListNode* n3 = new ListNode(3);
    ListNode* n4 = new ListNode(4);
    ListNode* n5 = new ListNode(5);
    n1->next = n2;
    n2->next = n3;
    n3->next = n4;
    n4->next = n5;
    
    int n = 2;
    
    Solution obj;
    ListNode* result = obj.removeNthFromEnd(n1, 2);
    return 0;
}

File: algorithm/go/firstround/19_2.go
package firstround

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil || n < 1 {
		return head
	}

	fast := head
	for i := 0; i < n; i++ {
		// 说明n>len(list)
		if fast == nil {
			return head
		}
		fast = fast.Next
	}

	dummy := &ListNode{Next: head}
	pre := dummy
	slow := head
	for fast != nil {
		pre = slow
		slow = slow.Next
		fast = fast.Next
	}

	pre.Next = slow.Next
	return dummy.Next
}
